### ISSUE TITLE
[mig modules] do not traverse non local file systems in filechecker (Bugzilla #896579)

### ISSUE THREAD

**Issue by jvehent** [mozilla/mig#92]
Migrated from https://bugzilla.mozilla.org/show_bug.cgi?id=896579
Assigned to: Julien Vehent [:ulfr]

On 2013-07-22 10:59:30 -0700, Julien Vehent [:ulfr] wrote:

> mounting points such as NFS and CIFS shouldn't be inspected. Or at least, there should be an option to prevent their inspection.

On 2014-05-25 10:31:51 -0700, Marco Vanotti wrote:

> This can be achieved using the statfs syscall[1] (syscall.Statfs in Go), and then checking the Type attribute from the result.
> However, should this be checked only against NFS and CIFS? (what about fuse? something mounted via sshfs should be avoided too?).
> 
> Is there a better way to do this?
> 
> 
> [1]: http://man7.org/linux/man-pages/man2/statfs.2.html

On 2014-05-30 23:56:16 -0700, Marco Vanotti wrote:

> An update for this issue:
> 
> In bsd, the statfs syscall gives you more info[1]: there's a f_flags field that will tell you if the filesystem is local (MNT_LOCAL). I think that will work for NFS/CIFS but reading freebsd's source code, it looks like FUSE filesystems are considered as local[2] (even if it's a sshfs filesystem).
> 
> In linux there's also a f_flag field in statfs, but it doesn't have the MNT_LOCAL flag.
> 
> FUSE filesystems don't report properly their f_type, so it will be hard to check. 
> 
> I'm not sure if the syscall is actually waiting for some response from the filesystem. At least with sshfs it is laggy.
> 
> Maybe the easiest way will be to parse /proc/mounts and check against a list of mounted fs. 
> 
> [1]: http://www.freebsd.org/cgi/man.cgi?query=statfs&sektion=2
> [2]: http://svnweb.freebsd.org/base/head/sys/fs/fuse/fuse_vfsops.c?revision=263233&view=markup#l339

On 2014-05-31 09:27:24 -0700, Julien Vehent [:ulfr] wrote:

> Thanks for the research Marco.
> I played with the statfs syscall in Go a little bit: https://gist.github.com/jvehent/8d86426a0b0918131410
> The above code is for Linux. It looks like building a blacklist of FS Types is the way to go.
> 
> About Fuse, I'm getting mixed results on Fedora 20. One path returns 0x27E0EB, another one 0x65735543. I'll need to look into it further to figure out why.

On 2014-05-31 12:23:07 -0700, Marco Vanotti wrote:

> Cool!
> 
> About FUSE: Both results look weird, in Ubuntu 13.10 (kernel 3.11) I'm getting: 0x65735546 which is also defined in [1]. Reading previous mailing list discussions[2] it looks like the only way to identify a fuse filesystem is by parsing /proc/mounts.
> 
> The problem with using statfs is that we are accessing the remote filesystem somehow. If there's an error with sshfs or nfs (for example, remote server disconnected), the program will wait forever. I think that's a problem that we should try to avoid.
> 
> If we parse /proc/mounts before starting we can get an idea of which filesystems are mounted and have a map[mountpoint]type to ask after accessing each folder. The problem with this is that it may happen that a remote fs can be mounted after parsing /proc/mounts.
> 
> Parsing /proc/mounts in Go: https://gist.github.com/mvanotti/66f23c32f1aa4cc2f042
> 
> 
> [1]: http://lxr.free-electrons.com/source/fs/fuse/inode.c#L50
> [2]: http://fuse.996288.n3.nabble.com/f-type-in-struct-statfs-td8823.html

**Comment by jvehent** [comment]
This hasn't been a problem for the last 2 years... I'm going to wontfix this, and we can always reopen if it becomes an issue.
